hazelcast/sql.py

```diff
@@ -0,0 +1,1604 @@
+          (iterator, update_count etc.) fails immediately
+        - setting it to an execute response
+
+        Args:
+            future (Future):
+        """
+        try:
+            response = future.result()
+
+            response_error = self._handle_response_error(response["error"])
+            if response_error:
+                # There is a server-side error sent to the client.
+                self._on_execute_error(response_error)
+                return
+
+            row_metadata = response["row_metadata"]
+            if row_metadata is not None:
+                # The result contains some rows, not an update count.
+                row_metadata = SqlRowMetadata(row_metadata)
+
+            self._on_execute_response(row_metadata, response["row_page"], response["update_count"])
+        except Exception as e:
+            # Something went bad, we couldn't get response from
+            # the server, invocation failed.
+            self._on_execute_error(self._sql_service.re_raise(e, self._connection))
+
+    def _handle_response_error(self, error):
+        """If the error is not ``None``, return it as
+        :class:`HazelcastSqlError` so that we can raise
+        it to user.
+
+        Args:
+            error (_SqlError): The error or ``None``.
+
+        Returns:
+            HazelcastSqlError: If the error is not ``None``,
+            ``None`` otherwise.
+        """
+        if error:
+            return HazelcastSqlError(error.originating_member_uuid, error.code, error.message, None)
+        return None
+
+    def _on_execute_error(self, error):
+        """Called when the first execute request is failed.
+
+        Args:
+            error (HazelcastSqlError): The wrapped error that can
+            be raised to the user.
+        """
+        with self._lock:
+            if self._closed:
+                # User might be already cancelled it.
+                return
+
+            self._execute_response.set_exception(error)
+
+    def _on_execute_response(self, row_metadata, row_page, update_count):
+        """Called when the first execute request is succeed.
+
+        Args:
+            row_metadata (SqlRowMetadata): The row metadata. Might be ``None``
+                if the response only contains the update count.
+            row_page (_SqlPage): The first page of the result. Might be
+                ``None`` if the response only contains the update count.
+            update_count (int): The update count.
+        """
+        with self._lock:
+            if self._closed:
+                # User might be already cancelled it.
+                return
+
+            if row_metadata:
+                # Result contains the row set for the query.
+                # Set the update count to -1.
+                response = _ExecuteResponse(row_metadata, row_page, -1)
+
+                if row_page.is_last:
+                    # This is the last page, close the result.
+                    self._mark_closed()
```

Comment:
`self._mark_closed()` also acquires the lock. How about using `self._closed = True` here?